### ISSUE TITLE
[Bugfix][DSA] DeepSeek-V4 case3 复读修复记录（A5/950PR）

### ISSUE THREAD

**Issue by QwertyJack** [GDzhu01/vllm-ascend-deepseekv4#108]
## 当前结论

case3 复读问题已在 A5 / 950PR + 最新正式权重下修复。

- 权重：`/home/weights/DeepSeek-V4-Flash`
- 硬件：A5 / 950PR，`ASCEND_RT_VISIBLE_DEVICES=6,7`
- 运行：DP=2，EP enabled，thinking enabled
- 修复提交：`bf9e4a3c`

## 根因与修复

1. `torch_npu.npu_transpose_quant_batchmatmul` 在 A5 / 950PR 上有精度问题。
   - 旧 `wo_a` 路径依赖该量化 batch matmul，长 thinking 下会放大小 logits 差异，导致错误分支持续自回归。
   - `vllm_ascend/quantization/methods/fp8.py:62-76`：`wo_a` 权重加载后反量化为 BF16。
   - `vllm_ascend/attention/dsa_v1.py:1405-1417`：BF16 `wo_a` 运行时走 `torch.matmul`，不再使用有精度风险的 transpose quant batch matmul。

2. `top_k=1` 遇到 tied / near-tied logits 时仍可能选错分支。
   - `vllm_ascend/sample/sampler.py:41-56`、`vllm_ascend/_310p/sample/sampler.py:51-54`：exact tie 只保留 deterministic argmax。
   - `vllm_ascend/worker/model_runner_v1.py:182`：定义 near-tie 阈值 `TOP_K_ONE_RESCORE_LOGIT_GAP = 0.25`。
   - `vllm_ascend/worker/model_runner_v1.py:473-532`：`top_k=1` 且 top1/top2 gap <= 0.25 时，用 FP32 hidden state dot LM-head candidate weights 对 top2 候选重打分，再 boost 选中的 token。
   - `vllm_ascend/quantization/methods/w8a8_mxfp8.py:39-75`：补 MXFP8 LM-head candidate rescore。

3. A5 DSA compressed-cache / sparse-attention 边界处理存在不稳定点。
   - `vllm_ascend/attention/dsa_v1.py:713-921`：prefill metadata 补真实 compressed seq lens。
   - `vllm_ascend/attention/dsa_v1.py:1680-1786`：A5 prefill compressor 按请求拆分，避免 padding / real compressed token 混写。
   - `vllm_ascend/attention/dsa_v1.py:1791-1947`：A5 prefill sparse attention 在多请求场景按请求拆分执行。
   - `vllm_ascend/core/single_type_kv_cache_manager.py:27-34,164-180,202-230`：compressed KV cache 按原始 block 粒度缓存/命中。
   - `vllm_ascend/patch/platform/patch_kv_cache_interface.py:128-233,248-348`、`vllm_ascend/worker/worker.py:544-546`、`vllm_ascend/worker/model_runner_v1.py:534-666`：补 grouped new-block tracking 和 KV cache zeroing。
   - `csrc/attention/compressor/op_kernel/arch35/compressor_kernel.h:359-363`、`csrc/attention/kv_compress_epilog/op_kernel/kv_compress_epilog_common.h:182-253,328-431,545-641`、`csrc/attention/kv_quant_sparse_attn_sharedkv/op_kernel/arch35/kv_quant_sparse_attn_sharedkv_scfa_block_vector.h:182,249-256,316`、`csrc/torch_binding.cpp:2405-2426`：修复自定义算子的空块、行偏移、稀疏 index、scale store 和输入检查。

4. DeepSeek-V4 DSA mixed prefill-decode PIECEWISE ACLGraph 在 A5 上仍不稳定。
   - 现象：显式 PIECEWISE 单请求可出现乱码循环，例如 `We trunk 4TOC7yNarius...`。
   - 策略：不声称 PIECEWISE 已修复；DeepSeek-V4 + A5 默认 ACLGraph 自动切到 `FULL_DECODE_ONLY`。
   - `vllm_ascend/platform.py:66-78`：识别 A5 + DeepSeek-V4。
   - `vllm_ascend/platform.py:353-360`：默认 `FULL_AND_PIECEWISE` 解析为 `FULL_DECODE_ONLY`。

### eager / no async / no prefix

```bash
/usr/bin/timeout 1200s python \
  $HOME/vllm/workspace/case3_concurrent_check_official.py \
  --tag case3_clean_conc8_full8192 --port 8778 --concurrency 8 \
  --max-tokens 8192 --source-port-start 52200
```

结果：

- `passed=true`
- 8/8 `finish_reason=stop`
- 8/8 最终答案 B
- 8/8 未检测到 repeat
- 输出自然结束，没有打满 8192

### ACLGraph default / FULL_DECODE_ONLY

启动脚本：

```bash
bash $HOME/vllm/workspace/run_case3_flash_aclgraph_official_noprefix_hccl300.sh
```

日志确认默认切到 `FULL_DECODE_ONLY`：

```text
Using FULL_DECODE_ONLY ACLGraph for DeepSeek-V4 on A5 because DSA mixed prefill-decode piecewise graph is not stable for this model.
FULL_DECODE_ONLY compilation enabled on NPU. use_inductor not supported - using only ACL Graph mode
```

最终复验命令：

```bash
/usr/bin/timeout 1200s python \
  $HOME/vllm/workspace/case3_concurrent_check_official.py \
  --tag case3_aclgraph_8e412b1c_conc8_full8192 --port 8778 \
  --concurrency 8 --max-tokens 8192 --source-port-start 53700
```

结果：

- `passed=true`
- 8/8 `finish_reason=stop`
- 8/8 最终答案 B
- 8/8 未检测到 repeat
- 输出 chunk 数：792 / 1170 / 1273 / 1559 / 1932，均自然 stop，没有 8192 runaway。
- summary：`/tmp/case3_aclgraph_8e412b1c_conc8_full8192_summary.json`

### 单测 / 静态检查

```bash
python -m py_compile \
  vllm_ascend/worker/model_runner_v1.py \
  vllm_ascend/platform.py \
  vllm_ascend/sample/sampler.py \
  vllm_ascend/_310p/sample/sampler.py \
  vllm_ascend/attention/dsa_v1.py \
  vllm_ascend/quantization/methods/fp8.py \
  vllm_ascend/quantization/methods/w8a8_mxfp8.py \
  vllm_ascend/core/recompute_scheduler.py \
  vllm_ascend/core/single_type_kv_cache_manager.py \
  vllm_ascend/models/deepseek_v4.py \
  vllm_ascend/patch/platform/patch_kv_cache_interface.py \
  vllm_ascend/worker/worker.py
```

结果：通过。

```bash
python -m pytest -q tests/ut/sample/test_sampler.py
```

结果：`2 passed`。

```bash
git diff --check
```

结果：通过。

未完成：

- 当前环境没有 `ruff`，所以 ruff check 未运行。

## 当前限制

- 不再声称 DSA mixed prefill-decode PIECEWISE ACLGraph 已修复；当前策略是在 DeepSeek-V4 + A5 默认使用 `FULL_DECODE_ONLY` ACLGraph。
- `torch_npu.npu_transpose_quant_batchmatmul` 不应作为 DeepSeek-V4 `wo_a` fallback，因为已确认在 A5 / 950PR 有精度问题。
